bokeh/sphinxext/bokehjs_content.py

```diff
@@ -0,0 +1,187 @@
+        if not hasattr(source_doc, 'bjs_seen'):
+            #we only want to inject the CODEPEN_INIT on one
+            #bokehjs-content block per page, here we check to see if
+            #bjs_seen exists, if not set it to true, and set
+            #node['include_bjs_header'] to true.  This way the
+            #CODEPEN_INIT is only injected once per document (html
+            #page)
+            source_doc.bjs_seen = True
+            node['include_bjs_header'] = True
+        cb = CodeBlock.run(self)
+        node.setup_child(cb[0])
+        node.children.append(cb[0])
+        return [target_node, node]
+
+def html_visit_bokehjs_content(self, node):
+
+    script_block = ""
+    for js_url in resources.js_files:
+        script_block += '''
+        <script type="text/javascript" src="%s"></script>
+        ''' % js_url
+
+    css_block = ""
+    for css_url in resources.css_files:
+        css_block += '''
+        <link rel="stylesheet" href="%s" type="text/css" />
+        ''' % css_url
+
+    if node['include_bjs_header']:
+        #we only want to inject the CODEPEN_INIT on one
+        #bokehjs-content block per page
+        self.body.append(BJS_CODEPEN_INIT.render(
+            css_block=css_block,
+            script_block=script_block))
+
+    self.body.append(
+        BJS_PROLOGUE.render(
+            id=node['target_id'],
+            title=node['title']))
+
+def html_depart_bokehjs_content(self, node):
+    self.body.append(BJS_EPILOGUE.render(
+        title=node['title']))
+
+def setup(app):
+    ''' Required Sphinx extension setup function. '''
+    app.add_node(
+        bokehjs_content,
+        html=(
+            html_visit_bokehjs_content,
+            html_depart_bokehjs_content
+        )
+    )
+    app.add_directive('bokehjs-content', BokehJSContent)
+
+#-----------------------------------------------------------------------------
+# Private API
+#-----------------------------------------------------------------------------
+
+#-----------------------------------------------------------------------------
+# Code
+#-----------------------------------------------------------------------------
+docs_cdn = settings.docs_cdn()
+
+# if BOKEH_DOCS_CDN is unset just use default CDN resources
+if docs_cdn is None:
+    resources = Resources(mode="cdn")
+else:
+    # "BOKEH_DOCS_CDN=local" is used for building and displaying the docs locally
+    if docs_cdn == "local":
+        resources = Resources(mode="server", root_url="/en/latest/")
+
+    # "BOKEH_DOCS_CDN=test:newthing" is used for building and deploying test docs to
+    # a one-off location "en/newthing" on the docs site
+    elif docs_cdn.startswith("test:"):
+        resources = Resources(
+            mode="server", root_url="/en/%s/" % docs_cdn.split(":")[1])
+
+    # Otherwise assume it is a dev/rc/full release version and use CDN for it
+    else:
+        resources = Resources(mode="cdn", version=docs_cdn)
+resources.js_components.append("bokeh-api")
```

Comment:
I think most of this can be removed, since the codepen will not function with anything other than CDN resources. I think you just need:
```
resources = Resources(mode="cdn")
resources.js_components.append("bokeh-api")
```
Things still won't work on CDN with e.g. local resources, but they wouldn't anyway, so we might as well keep this simple.

Reply:
actually, `BOKEH_DOCS_CDN=<real version>` can work too, so then:
```
docs_cdn = settings.docs_cdn()
if docs_cdn is None:
    resources = Resources(mode="cdn")
else:
    resources = Resources(mode="cdn", version=docs_cdn)
resources.js_components.append("bokeh-api")
```

Reply:
Ah, except that will fail with `BOKEH_DOCS_CDN="local"` since a version validation will throw an exception. OK, so then let's keep the logic as is, but not duplicate in two places. This shoudl be factored out into local `util.py ` or something